Add linked-on-or-after checks for network process blob file URL checking

achristensen07 opened the pull request:
#### c9b868dd9fb721812f0916118270135f4dde6656
<pre>
Add linked-on-or-after checks for network process blob file URL checking
<a href="https://bugs.webkit.org/show_bug.cgi?id=281798">https://bugs.webkit.org/show_bug.cgi?id=281798</a>
<a href="https://rdar.apple.com/137326900">rdar://137326900</a>

Reviewed by Brent Fulgham.

281966@main introduced checks in the network process that verify blob URLs are allowed to be
read by a certain web process before reading them.  This required introducing IPC roundtrips
between the UI process and network process.  However, the introduction of these IPC roundtrips
subtly changed the callback sequence, and apps that rapidly call WKWebView.loadData, loadFileURL,
and loadRequest without waiting for them to finish and responding to WKNavigationDelegate&apos;s
decidePolicyForNavigationAction and decidePolicyForNavigationResponse with ...PolicyCancel
based on the callback sequence now sometimes cancel the wrong load.  To keep these apps working
until they update, I introduce a linked-on-or-after check for the IPC roundtrips and the
check in the network process they facilitate.

* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::shouldCheckBlobFileAccess):
(WebKit::NetworkConnectionToWebProcess::registerInternalFileBlobURL):
(WebKit::NetworkConnectionToWebProcess::registerInternalBlobURLOptionallyFileBacked):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::networkProcessWillCheckBlobFileAccess):
(WebKit::WebProcessProxy::assumeReadAccessToBaseURL):
(WebKit::WebProcessProxy::assumeReadAccessToBaseURLs):

Canonical link: <a href="https://commits.webkit.org/285537@main">https://commits.webkit.org/285537@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ea45cf0f66dc55a362659b3c9dadd5f6be364af

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72732 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52157 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25531 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76923 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23966 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59962 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23794 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57185 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15683 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75799 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47163 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62598 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37613 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43811 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20064 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22295 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/65865 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65665 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20423 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78598 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/71988 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16978 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19581 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65639 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/17026 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62606 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64916 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13221 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6868 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/93767 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11230 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47955 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2742 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20629 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/49022 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50317 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48767 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -960,6 +969,10 @@ void WebProcessProxy::assumeReadAccessToBaseURL(WebPageProxy& page, const String
         completionHandler();
     };
+
+    if (!networkProcessWillCheckBlobFileAccess())
```
Nit: Should we NSLog something here to let the racy application author know they need to handle async calls properly?

brentfulgham replied:
On second thought, since this is just a LOA check, it's not very useful to tell them they aren't getting the blob access check behavior.

achristensen07 replied:
Yeah in this case it's pretty obvious.  They'll get callbacks in a different order, their application will break, and they'll fix it.

brentfulgham reviewed (approved):
I suggest logging something when you do this, but I guess since it's just a linked-on-or-after (and not a bundle check) it's probably not that necessary.

achristensen07 committed:
Add linked-on-or-after checks for network process blob file URL checking
https://bugs.webkit.org/show_bug.cgi?id=281798
rdar://137326900

Reviewed by Brent Fulgham.

281966@main introduced checks in the network process that verify blob URLs are allowed to be
read by a certain web process before reading them.  This required introducing IPC roundtrips
between the UI process and network process.  However, the introduction of these IPC roundtrips
subtly changed the callback sequence, and apps that rapidly call WKWebView.loadData, loadFileURL,
and loadRequest without waiting for them to finish and responding to WKNavigationDelegate's
decidePolicyForNavigationAction and decidePolicyForNavigationResponse with ...PolicyCancel
based on the callback sequence now sometimes cancel the wrong load.  To keep these apps working
until they update, I introduce a linked-on-or-after check for the IPC roundtrips and the
check in the network process they facilitate.

* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::shouldCheckBlobFileAccess):
(WebKit::NetworkConnectionToWebProcess::registerInternalFileBlobURL):
(WebKit::NetworkConnectionToWebProcess::registerInternalBlobURLOptionallyFileBacked):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::networkProcessWillCheckBlobFileAccess):
(WebKit::WebProcessProxy::assumeReadAccessToBaseURL):
(WebKit::WebProcessProxy::assumeReadAccessToBaseURLs):

Canonical link: https://commits.webkit.org/285537@main
